### PR TITLE
CHE-5725: Run all java tests from the project or package

### PR DESCRIPTION
**Pull request by svor**
### What does this PR do?

- Run java tests (JUnit4, TestNg) from the Project Explorer
  - Can run one test class

  - Can run tests from package
package
  - Can run tests from project
project

For this case (when the editor isn't active) we have  actions:
selection_001
selection_003

- Run java test from the editor. When the editor is active , the test runner will automatically detect test framework. For this case we have next set of actions:
selection_002

### What issues does this PR fix or reference?
https://github.com/eclipse/che/issues/5725
https://github.com/eclipse/che/issues/5728

#### Changelog

Run all java tests from the project or package

#### Release Notes

Run all java tests from the project or package

### CODE REVIEW

**Review by davidfestal — COMMENTED** [review]
Nice, but it seems it still doesn't manage the on-the-fly compilation of projects before the first test run ?

**Comment by svor** [comment]
Still don't have project compilation before test running. 
Have an issue for this (https://github.com/eclipse/che/issues/5422)

File: plugins/plugin-testing-java/plugin-testing-junit/che-plugin-testing-junit-ide/src/main/java/org/eclipse/che/plugin/testing/junit/ide/action/JUnitAbstractAction.java [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+/**
+ * Abstract action for JUnit test actions.
+ */
+public abstract class JUnitAbstractAction extends RunDebugTestAbstractAction {
+    public static final String JUNIT_TEST_FRAMEWORK = "junit";
+
+    private AppContext appContext;
+    private boolean    isEnable;
+
+    protected TestExecutionContext.TestType testType;
```
Comment (vparfonov) [inline plugins/plugin-testing-java/plugin-testing-junit/che-plugin-testing-junit-ide/src/main/java/org/eclipse/che/plugin/testing/junit/ide/action/JUnitAbstractAction.java]: TestType not good name it think, as i understand its context for runinig

File: plugins/plugin-testing-java/plugin-testing-junit/che-plugin-testing-junit-ide/src/main/java/org/eclipse/che/plugin/testing/junit/ide/action/RunJUnitTestAction.java [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+        notification.setStatus(SUCCESS);
+        if (result.isSuccess()) {
+            notification.setTitle("Test runner executed successfully.");
+            testResultPresenter.handleResponse();
+        } else {
+            notification.setTitle("Test runner failed to execute.");
+        }
+    }
+
+    private void onTestRanUnsuccessfully(JsonRpcError exception, StatusNotification notification) {
```
Comment (vparfonov) [inline plugins/plugin-testing-java/plugin-testing-junit/che-plugin-testing-junit-ide/src/main/java/org/eclipse/che/plugin/testing/junit/ide/action/RunJUnitTestAction.java]: onTestFail() ?

**Comment by MatousJobanek** [comment]
Hi,
are the private methods in [TestingRPCService](https://github.com/eclipse/che/blob/master/wsagent/che-core-api-testing/src/main/java/org/eclipse/che/api/testing/server/TestingRPCService.java) still somewhere used? They seem to be like a zombie for me. The same for the method [AbstractJavaTestRunner#detectTests](https://github.com/eclipse/che/blob/master/plugins/plugin-testing-java/plugin-testing-classpath/src/main/java/org/eclipse/che/plugin/java/testing/AbstractJavaTestRunner.java#L58) (and the related methods) as you are using a new approach for retrieving classes, right?

**Comment by svor** [comment]
Hi @MatousJobanek 
About `TestingRPCService` - it is a service which uses JSON RPC concepts. It register two methods (for detecting and running tests). It means if the client will send a request via websocket with name `RUN_TESTS_METHOD` - server will call method `org.eclipse.che.api.testing.server.TestingRPCService#runTests` the same about `TESTING_RPC_TEST_DETECTION_NAME` and `org.eclipse.che.api.testing.server.TestingRPCService#handleTestDetection`. 

 About `AbstractJavaTestRunner#detectTests` - this method analyzes opened test file and detects tests methods, and which framework we should use. This detection needs only for analyzing opened editor.

File: plugins/plugin-testing-java/plugin-testing-junit/che-plugin-testing-junit-ide/src/main/java/org/eclipse/che/plugin/testing/junit/ide/action/JUnitAbstractAction.java (hunk 1 above)
Comment (svor) [reply plugins/plugin-testing-java/plugin-testing-junit/che-plugin-testing-junit-ide/src/main/java/org/eclipse/che/plugin/testing/junit/ide/action/JUnitAbstractAction.java]: Changed to ContextType

File: plugins/plugin-testing-java/plugin-testing-junit/che-plugin-testing-junit-ide/src/main/java/org/eclipse/che/plugin/testing/junit/ide/action/RunJUnitTestAction.java (hunk 2 above)
Comment (svor) [reply plugins/plugin-testing-java/plugin-testing-junit/che-plugin-testing-junit-ide/src/main/java/org/eclipse/che/plugin/testing/junit/ide/action/RunJUnitTestAction.java]: It isn't exactly test fail. Renamed to `onTestRanFailed`

**Comment by MatousJobanek** [comment]
Ah, I missed that. Thx for clarification @svor

**Comment by MatousJobanek** [comment]
Have you think about extracting the code from `*TestFinder` classes (`TestNGTestFinder` & `JUnit4TestFinder`) to some shared module - most of the code seems to be same - you know, to bring some higher abstraction. Similar also for `createSuite` method - move most of the code to `AbstractJavaTestRunner` and then just providing a specific implementation of some method would define how the classes will be detected.

**Comment by svor** [comment]
Make sense, I will do it. Thx!

**Review by vparfonov — APPROVED** [review]

**Comment by svor** [comment]
ci-test

File: plugins/plugin-testing/che-plugin-testing-ide/src/main/java/org/eclipse/che/plugin/testing/ide/action/RunDebugTestAbstractAction.java [hunk 3]
```diff
@@ -34,55 +35,59 @@
...
-                               SVGResource icon) {
+    private TextEditor          currentEditor;
+    private List<TestPosition>  testPosition;
+    private TestServiceClient   client;
+    private DtoFactory          dtoFactory;
+    private AppContext          appContext;
+    private NotificationManager notificationManager;
+
+    public boolean isEditorInFocus;
+    public boolean isEnable;
```
Comment (MatousJobanek) [inline plugins/plugin-testing/che-plugin-testing-ide/src/main/java/org/eclipse/che/plugin/testing/ide/action/RunDebugTestAbstractAction.java]: I just came across this part - is it really necessary to have these variables `public` when they are not constants?

**Review by MatousJobanek — COMMENTED** [review]
Thank you for refactoring the code @svor This is exactly what I was expecting. Apart from the one comment inside of the code I have another proposal for refactoring/improvement. In last two days, I was playing with the IDE side and I noticed that the classes `JUnitAbstractAction` and `TestNGAbstractAction` are completely same. The same I can say about classes `Run*TestAction` and `Debug*TestAction`. Do you have any plan for merging the classes together?

**Comment by svor** [comment]
@MatousJobanek Thank you for reviewing! I've made some optimizations with actions and I think now it's more clear.

**Review by evidolob — APPROVED** [review]

**Pull request merged**